### PR TITLE
ADR 005: update client consensus height events 

### PR DESCRIPTION
**Pull request by damiannolan**
## Description

- Adding ADR 005 - `UpdateClient Events - ClientState Consensus Heights`

closes: #1280 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

**Comment by damiannolan** [comment]
All suggestions welcome, originally I had written more but decided to trim it back to keep this concise.

File: docs/architecture/adr-005-consensus-height-events.md [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+* 25/04/2022: initial draft
+
+## Status
+
+Accepted 
+
+## Context
+
+The `ibc-go` implementation leverages the [Cosmos-SDK's EventManager](https://github.com/cosmos/cosmos-sdk/blob/main/docs/core/events.md#EventManager) to provide subscribers a method of reacting to application specific events.
+IBC relayers depend on the [`consensus_height`](https://github.com/cosmos/ibc-go/blob/main/modules/core/02-client/keeper/events.go#L33) attribute emitted as part of `UpdateClient` events in order to run misbehaviour detection by cross-checking the details of the *Header* emitted at a given consensus height against that of the *Header* from the originating chain.
```
Comment (crodriguezvega) [inline docs/architecture/adr-005-consensus-height-events.md:13]: I guess you mean that it cross-checks the details... Would you be able to specify in more details what those details are?

```suggestion
IBC relayers depend on the [`consensus_height`](https://github.com/cosmos/ibc-go/blob/main/modules/core/02-client/keeper/events.go#L33) attribute emitted as part of `UpdateClient` events in order to run misbehaviour detection by cross-checking the details of the *Header* emitted at a given consensus height against those of the *Header* from the originating chain.
```

File: docs/architecture/adr-005-consensus-height-events.md [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+Following the refactor of the `02-client` submodule and associated `ClientState` interfaces, it will now be possible for
+light client implementations to perform such actions as batch updates, inserting `N` number of `ConsensusState`s into the application state tree with a single `UpdateClient` message. This flexibility is provided in `ibc-go` by the usage of the [Protobuf Any](https://developers.google.com/protocol-buffers/docs/proto3#any) field contained within the [`UpdateClient`](https://github.com/cosmos/ibc-go/blob/main/proto/ibc/core/client/v1/tx.proto#L44) message.
```
Comment (crodriguezvega) [inline docs/architecture/adr-005-consensus-height-events.md:16]:
```suggestion
light client implementations to perform such actions as batch updates, inserting `N` number of `ConsensusState`s into the application state tree with a single `UpdateClient` message. This flexibility is provided in `ibc-go` by the usage of the [Protobuf `Any`](https://developers.google.com/protocol-buffers/docs/proto3#any) field contained within the [`UpdateClient`](https://github.com/cosmos/ibc-go/blob/main/proto/ibc/core/client/v1/tx.proto#L44) message.
```

File: docs/architecture/adr-005-consensus-height-events.md [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+	ClientType() string
+	ValidateBasic() error
+}
+```
+
+To support this functionality the `GetHeight()` method has been omitted from the new `ClientMessage` interface.
+Emission of standardised events from the `02-client` submodule now becomes problematic and is two-fold:
+
+1. The `02-client` submodule previously depended upon the `GetHeight()` method of `Header` types in order to [retrieve the updated consensus height](https://github.com/cosmos/ibc-go/blob/main/modules/core/02-client/keeper/client.go#L90).
```
Comment (crodriguezvega) [inline docs/architecture/adr-005-consensus-height-events.md:42]: This link to https://github.com/cosmos/ibc-go/blob/main/modules/core/02-client/keeper/client.go#L90 will become unusable was we merge the 02-client refactor to main, right? Maybe better to link a specific tag?

File: docs/architecture/adr-005-consensus-height-events.md [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+2. Emitting a single `consensus_height` event attribute is not sufficient in the case of a batched client update containing multiple *Headers*.
+
+## Decision
+
+The following decisions have been made in order to provide flexibility to consumers of `UpdateClient` events in a non-breaking fashion:
+
+1. Maintain the `consensus_height` event attribute emitted from the `02-client` update handler, but mark as deprecated for future removal. For example, with tendermint lightclients this will simply be `consensusHeights[0]` following a successful update using a single *Header*.
```
Comment (crodriguezvega) [inline docs/architecture/adr-005-consensus-height-events.md:49]: You mention here `consensusHeights[0]` but I miss a bit of context of where it comes from...

File: docs/architecture/adr-005-consensus-height-events.md (hunk 1 above)
Comment (crodriguezvega) [inline docs/architecture/adr-005-consensus-height-events.md:13]: Maybe also replace this link to main https://github.com/cosmos/ibc-go/blob/main/modules/core/02-client/keeper/events.go#L33 with a tag.

**Review by crodriguezvega — COMMENTED** [review]
Your first ADR, yuhhuu! 🥳

File: docs/architecture/adr-005-consensus-height-events.md (hunk 1 above)
Comment (damiannolan) [reply docs/architecture/adr-005-consensus-height-events.md:13]: Sure, I really paraphrased this part from https://github.com/cosmos/ibc-go/pull/1208#discussion_r851167308

But I will try to add more details here, although I think this is technically hermes specific behaviour and also required the consumer to be able to parse and unmarshal the `ClientMessage`(`Header`) from the events so perhaps it should read "Some IBC relayers depend on the consensus height...."

File: docs/architecture/adr-005-consensus-height-events.md (hunk 4 above)
Comment (damiannolan) [reply docs/architecture/adr-005-consensus-height-events.md:49]: Good spot, I think this would make more sense with point 2. as point 1. and vice versa

Comment (damiannolan) [reply docs/architecture/adr-005-consensus-height-events.md]: Updated and switched ordering 👍

File: docs/architecture/adr-005-consensus-height-events.md (hunk 1 above)
Comment (damiannolan) [reply docs/architecture/adr-005-consensus-height-events.md]: Updated with additional info using [this source](https://github.com/informalsystems/ibc-rs/blob/master/modules/src/clients/ics07_tendermint/header.rs#L23)

**Review by colin-axner — APPROVED** [review]
Outstanding. Well written :100:

**Review by crodriguezvega — APPROVED** [review]

**Pull request merged**